### ISSUE TITLE
imitate leak

### ISSUE THREAD

**Issue by whitecolor** [staltz/xstream#51]
This `imitate` code seem to leak, doesn't it?
```js
import xs from 'xstream'
import delay from 'xstream/extra/delay'

let proxy$ = new xs.createMimic()

let source$ = proxy$.startWith(1).map(x => x*2)
  .debug(x => console.log('source$ tap', x))  

let result$ = source$.compose(delay(1000))

proxy$.imitate(result$)

result$
  .take(4)
  .addListener({
    next: x => console.log('x', x),
    error: () => {},
    complete: () => {}
  })
```

```
source$ tap 2
x 2
source$ tap 4
x 4
source$ tap 8
x 8
source$ tap 16
x 16
source$ tap 32 // here a leak starts
source$ tap 64
source$ tap 128
....
```
http://www.webpackbin.com/E17QfBUVb

I was pretty sure that `imitate` should protect from such leak that was (actually is) a problem with rx proxy subjects.

**Comment by staltz** [comment]
Thanks for the report. We'll do something about this soon enough.

**Comment by whitecolor** [comment]
I'm wondering how are you going to solve this. Also interested in proper solution for rx (https://github.com/Reactive-Extensions/RxJS/issues/1259) and actually @TylorS's https://github.com/TylorS/most-proxy has the same issue.

**Comment by staltz** [comment]
@whitecolor I think this issue is already solved. I wrote this test, and it passes the current version (4.0.4):

```js

  it('should not create leaked cyclic executions', (done) => {
    const expectedMimic = [2, 4, 8, 16];
    const expectedResult = [2, 4, 8, 16];

    const proxy$ = xs.createMimic<number>();
    const source$ = proxy$.startWith(1).map(x => x * 2)
      .debug(x => {
        assert.equal(expectedMimic.length > 0, true);
        assert.equal(x, expectedMimic.shift());
      });
    const result$ = source$.compose(delay(100)).compose(s => <Stream<number>> s);
    proxy$.imitate(result$)

    result$.take(4).addListener({
      next: (x: number) => {
        assert.equal(x, expectedResult.shift());
      },
      error: (err: any) => done(err),
      complete: () => {
        assert.equal(expectedMimic.length, 0);
        assert.equal(expectedResult.length, 0);
        setTimeout(() => {
          done();
        }, 1000);
      },
    });
  });
```

**Comment by whitecolor** [comment]
But did you check this http://www.webpackbin.com/E17QfBUVb example? (was incorrect link in OP)

**Comment by staltz** [comment]
Yes I did, but it's basically the same code as the test.

**Comment by whitecolor** [comment]
@staltz are you sure that your test is correct? ) you probably should use `.push` to increase the number of produced elements (by `listener` and by `debug` operator), and not `shit` applied to empty array million times that will give the same result.

Here is this code:
http://www.webpackbin.com/4yiYcHLNZ

**Comment by staltz** [comment]
Ok, it was a problem with mocha.
This test reproduces the bug:

```js

  it('should not create leaked cyclic executions', (done) => {
    const expectedMimic = [2, 4, 8, 16];
    const expectedResult = [2, 4, 8, 16];

    const proxy$ = xs.createMimic<number>();
    const source$ = proxy$.startWith(1).map(x => x * 2)
      .debug(x => {
        try {
          assert.equal(expectedMimic.length > 0, true,
            'should be expecting some next value');
          assert.equal(x, expectedMimic.shift());
        } catch (err) {
          done(err);
        }
      });
    const result$ = source$.compose(delay(100)).compose(s => <Stream<number>> s);
    proxy$.imitate(result$)

    result$.take(4).addListener({
      next: (x: number) => {
        assert.equal(x, expectedResult.shift());
      },
      error: (err: any) => done(err),
      complete: () => {
        assert.equal(expectedMimic.length, 0);
        assert.equal(expectedResult.length, 0);
        setTimeout(() => {
          done();
        }, 1000);
      },
    });
  });
```

Thanks.

**Comment by whitecolor** [comment]
@staltz what is your thoughts, how this can be solved conceptually?

**Comment by staltz** [comment]
@whitecolor https://github.com/staltz/xstream/commit/8a432b66523924b48adcd974619d0baa98c4f133

This is now fixed. Releasing v4.0.5 soon.

**Comment by whitecolor** [comment]
Cool, this kind of what I was thinking too, it should be "hidden listeners".

**Comment by staltz** [comment]
Actually, I won't release this in 4.0.5, I'll release it in v5.0.0, because I'll move imitate from MimicStream to Stream, then delete MimicStream and createMimic().

**Comment by whitecolor** [comment]
You major releasing is very consistent) seriously I like it.

**Comment by whitecolor** [comment]
@staltz will you test work if to replace:
```
 result$.take(4).addListener({
```
with
```
source$.take(4).addListener({
```
because this will be **more proper test** as stream that we want to mimic (`result$`) may not have existing subscription in the app, but mimic should work still.

**Comment by whitecolor** [comment]
@staltz will you test work if to replace:
```js
 result$.take(4).addListener({
```
with
```js
source$.take(4).addListener({
```
because this will be **more proper test** as stream that we want to mimic (`result$`) may not have existing subscriptions in the app, but mimic should work still (naming in the test should be made more appropriate probably to this case).

**Comment by Hypnosphi** [comment]
It is quite natural to expect here that `results$` here gets all the subscriptions attached to `proxy$` and loses them as soon as they are removed

**Comment by whitecolor** [comment]
@staltz it leaks in 5.0.3
http://www.webpackbin.com/NyZXA0v4b (see the console)

**Comment by Hypnosphi** [comment]
Basically, the problem here is that whenever we have a cycle, there is some stream which gets two internal listeners triggered by the same endpoint subscriber. So when subscriber detaches, this node still has one `il`. Maybe if `ils` could have a reference to their initiators (endpoint subscribers), it would allow to prevent such duplication.

**Comment by staltz** [comment]
I think I just need to build a cycle detection system to cleanup those cycles. Seems unavoidable.

**Comment by whitecolor** [comment]
Just my thoughts on the topic, I don't know how subscription chain system works in xstream.

The problem is when `mimicked` (`orignal`) stream has dependency on `mimic`. When `mimic` initiated (with `imitate`) it creates some subscriptions, we should end this subscriptions. at a certain moment, and the problem is to define when this moment happens, Basically it is the moment when `original` is only subscribed by chain that was stemmed by `mimic` subscription. If you can detect moment when `orginal` subscription chain has only `mimic` has as source `mimic` can be unsubscribed.

**Comment by staltz** [comment]
@whitecolor fixed: https://github.com/staltz/xstream/commit/8658aa0c8f1a5e72d638306741c67a163dcb523f

**Comment by whitecolor** [comment]
At least test does pass. Btw I don't see a reason why it should not work memory stream.

This phrase from FAQ:
> It's useful to think about MemoryStreams as cells in a spreadsheet. It doesn't make any sense to define a spreadsheet cell A1 with a formula that depends on B1 and cell B1 defined with a formula that depends on A1.

This is true for sync cycle (and cell in a spreadsheet is sync usually), but seem not to be true for async cycle. Maybe I missing something.

**Comment by staltz** [comment]
And MemoryStreams have sync emission built-in, because a listener is added, an `next` is sent synchronously. See it now?

**Comment by whitecolor** [comment]
> an `next` is sent synchronously

What is "`next`"? Why this causes a problem? Can you show the code that identifies the issue with memory stream?

**Comment by whitecolor** [comment]
@staltz 
can't get a little: if we in the test move `imitate` after `addListener` it won't work. Is it expected?

**Comment by whitecolor** [comment]
@staltz I believe I found another bug: seem because of the guy with `take(1)` in the middle.
http://www.webpackbin.com/EkU7JrFNW

Look at the console.

**Comment by Hypnosphi** [comment]
Basically, just 2 listeners are enough to reproduce, if the second detaches itself before first: http://www.webpackbin.com/VJ0D-9YVW

**Comment by staltz** [comment]
Can you open a new issue, please? These aren't anymore about leaks.

**Comment by whitecolor** [comment]
@Hypnosphi I see that you example works ok, no any leak happens after all listeners detached.

@staltz ok I will create another issue but it is still about kinda leak
